### PR TITLE
kernel/pthread: make internal pthread_join function to remove duplication

### PR DESCRIPTION
**Pull request by sunghan-chang**
The pthread_join and pthread_tryjoin_np are almost the same except
blocking. It causes many codes duplication.
This commit adds pthread_join_internal function and
pthread_join and pthread_tryjoin_np calls it with blocking option.

It already gets rtcb through this_task() so that it does not
need to call getpid for pid value.
Let's get it directly from rtcb structure.

### CODE REVIEW

**Comment by seinfra** [comment]
Target : [9431a6b19ab1f2eada90a0d840aaa234cd2d5c24] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [9431a6b19ab1f2eada90a0d840aaa234cd2d5c24] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [91a6a38ca91ce6547f12c634d8ee318056fe1544] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [91a6a38ca91ce6547f12c634d8ee318056fe1544] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [10be6ba9e2b16b555a8d5457c5641a27ebf8cd2e] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [10be6ba9e2b16b555a8d5457c5641a27ebf8cd2e] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [dd14b7de5cd6e5359652b7074447361edab0873a] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [dd14b7de5cd6e5359652b7074447361edab0873a] - Code Rule Check OK.

File: apps/examples/testcase/le_tc/kernel/tc_pthread.c [hunk 1]
```diff
@@ -176,6 +185,21 @@ static void *pthread_exit_thread(void *param)
...
+/**
+* @fn                   :self_pthread_join_n_exit
+* @brief                :utility function for tc_pthread_pthread_create_exit_join
+* @return               :void *
+*/
+static void *self_pthread_join_n_exit(void *param)
+{
+	pid_t pid = getpid();
+
+	chk_val = pthread_join((pthread_t)pid, NULL);
```
Comment (sangam-swami) [inline apps/examples/testcase/le_tc/kernel/tc_pthread.c]: do we need to check return value of chk_val

**Comment by sangam-swami** [comment]
The pthread_join and pthread_tryjoin_np are almost the same except
pthread_tryjoin_np is blocking call. It causes many codes duplication.
This commit adds pthread_join_internal function.
pthread_join and pthread_tryjoin_np(with blocking option) calls it as common function.

File: apps/examples/testcase/le_tc/kernel/tc_pthread.c (hunk 1 above)
Comment (sunghan-chang) [reply apps/examples/testcase/le_tc/kernel/tc_pthread.c]: This case is checking "attempt to join to itself (deadlock)". 
```
int pthread_join_internal(pthread_t thread, FAR pthread_addr_t *pexit_value, bool blocking)
{
	... <snip> ...
	if ((pid_t)thread == rtcb->pid) {
		return EDEADLK;
	}
	... <snip> ...
}
```
In this case, checking return value is necessary.

**Comment by sunghan-chang** [comment]
> The pthread_join and pthread_tryjoin_np are almost the same except
pthread_tryjoin_np is blocking call. It causes many codes duplication.
This commit adds pthread_join_internal function.
pthread_join and pthread_tryjoin_np(with blocking option) calls it as common function.

I rewrite as below. Thank you for helping.
```
The pthread_join and pthread_tryjoin_np are almost the same
except pthread_tryjoin_np is non-blocking call.
This commit adds pthread_join_internal function to resolve code duplication.
pthread_join (with blocking option) and pthread_tryjoin_np (with non-blocking option)
calls it as common function.
```

**Comment by seinfra** [comment]
Target : [11fbf47f7f98f93660885adabf0053b9b2ff570b] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [11fbf47f7f98f93660885adabf0053b9b2ff570b] - Code Rule Check OK.

**Pull request merged**